include/gmsh.tcc

```diff
@@ -194,12 +194,111 @@ void GMSH<Tdim, Tvertices>::store_element_vertices() {
+
+  const unsigned Tdim = 2;
+  const unsigned Tvertices = 4;
+
+  //! Storing ngauss_points to member variable and get constants from namespace
+  ngauss_points_ = ngauss_points;
+  std::vector<double> gauss_constants =
+      element::gauss_points.find(ngauss_points_)->second;
+
+  //! Create a matrix of xi from gauss points
+  //! Matrix is size npoints x Tdim
+  //! For 2D only
+  unsigned npoints = std::pow(ngauss_points_, Tdim);
+  Eigen::MatrixXd xi_gauss_points(npoints, Tdim);
+  unsigned counter = 0;
+  for (unsigned ii = 0; ii < ngauss_points_; ++ii) {
+    for (unsigned jj = 0; jj < ngauss_points_; ++jj) {
+        xi_gauss_points(counter, 0) = gauss_constants.at(ii);
+        xi_gauss_points(counter, 1) = gauss_constants.at(jj);
+        ++counter;
+    }
+  }
+
+  Eigen::VectorXd pointsarray(Tdim);
+
+  //! global_id is the index for all points
+  unsigned global_id = 0;
+  //! material_id is the index of materialpoints
+  unsigned material_id = 0;
+  //! point_id is local index of points
+  unsigned point_id = 0;
+
+  //! Update vector of material points
+  //! Fill materialpoints_ vector for the first component
+  materialpoints_.emplace_back(std::shared_ptr<MaterialPoints<Tdim>>(
+      new MaterialPoints<Tdim>(material_id)));
+  for (const auto& element : elements_) {
+
+    //! Store coordinates in Tdim x Tvertices matrix
+    //! Where N is the number of nodes per element
+    //! This is rearranging of the data to have stored in matrix form
+    Eigen::MatrixXd node_coordinates(Tdim, Tvertices);
+    for (unsigned i = 0; i < Tvertices; ++i) {
+      for (unsigned j = 0; j < Tdim; ++j) {
+        node_coordinates(j, i) = element->vertex_coordinates(i)[j];
+      }
+    }
+
+    //! Compute the volume for each point
+    //! The volume of the element is distributed evenly
+    double point_volume = element->calculate_volume() / npoints;
+
+    //! Compute the gauss points (there are npoints)
+    for (unsigned k = 0; k < npoints; ++k) {
+
+      //! Get array of xi for this gauss point
+      std::array<double, Tdim> xi;
+      for (unsigned l = 0; l < Tdim; ++l) xi.at(l) = xi_gauss_points(k, l);
+
+      //  gauss point in cartesian coordinate
+      for (unsigned i = 0; i < Tdim; ++i) {
+        pointsarray[i] = 0;
+        
+        Eigen::VectorXd shape_function = element::quadrilateral::shapefn(xi);
+
+        for (unsigned j = 0; j < Tvertices; ++j) {
+          pointsarray[i] += shape_function[j] * node_coordinates(i, j);
+        }
+      }
+
+      //! Make class point and store to material points
+      std::shared_ptr<Point<Tdim>> point = std::make_shared<Point<Tdim>>(
+          point_id, global_id, pointsarray, point_volume);
+      materialpoints_.at(material_id)->add_points(point);
+
+      //! Update point_id
+      ++point_id;
+      //! Update global_id
+      global_id = point_id;
+    }
+  }
+
+  //! Find number of material points generated
+  for (const auto& materialpoint : materialpoints_)
```

Comment:
This naming is unclear, what is `materialpoint`?

Reply:
isn't this quite typical for range-for loop?

Reply:
`materialpoints_` is a collection of material points, so I can see why it has `npoints`, but I don't understand why a single `materialpoint` has `npoints`

Reply:
Could you iterate from `materialpoints : materialpoints_` it's not a singular material point